feat: add `utils/some-own-by`

performant23 opened the pull request:
Add a utility to test whether some own properties of a provided object satisfies a predicate function

Fixes: #821

Resolves #821.

## Description

> What is the purpose of this pull request?
Add a utility to test whether some own properties of a provided object satisfies a predicate function

This pull request:

-  The function  implementation works by iterating over the “own” properties of an object and applying a predicate function to each property.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #821
-   fixes #821

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

performant23 committed:
feat(@stdlib/utils): if applied, this commit will add 'someOwnBy' in utils

Add a utility to test whether some own properties of a provided object satisfies a predicate function

Fixes: #821

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/README.md:
```diff
@@ -0,0 +1,238 @@
+
+#### someOwnBy( obj, n, predicate\[, thisArg ] )
+
+Tests whether an obj contains at least n own properties which pass a test implemented by a predicate function.
```
```suggestion
Tests whether an `object` contains at least `n` own properties which pass a test implemented by a `predicate` function.
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/README.md:
````diff
@@ -0,0 +1,238 @@
+// returns true
+```
+
+Once the function finds n successful properties, the function immediately returns true.
````
```suggestion
Once the function finds `n` successful properties, the function **immediately** returns `true`.
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/README.md:
````diff
@@ -0,0 +1,238 @@
+// returns true
+```
+
+The invoked function is provided three arguments:
````
```suggestion
The invoked `function` is provided three arguments:
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/README.md:
```diff
@@ -0,0 +1,238 @@
+
+The invoked function is provided three arguments:
+
+-   `value:` object property value
+-   `key`: object property key
+-   `obj`: input object
```
```suggestion
-   `value`: object property value
-   `key`: object property key
-   `obj`: input object
```


Not sure if this is correct, please refer nearby packages.

performant23 replied:
Yeah, you're correct! I just mistyped.

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/README.md:
```diff
@@ -0,0 +1,238 @@
+
+-   An [`Object`][mdn-object] refers to a JavaScript object, which is a collection of properties. Each property is an association between a key (or name) and a value. The key can be a string or a symbol, and the value can be any JavaScript value, including functions and other objects
+
+-   If provided an empty obj, the function returns false.
```
```suggestion
-   If provided an empty `object`, the function returns `false`.
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/README.md:
````diff
@@ -0,0 +1,238 @@
+    */
+    ```
+
+-   The function provides limited support for dynamic collections (i.e., collections whose `length` changes during execution).
````
```suggestion
-   The function provides limited support for dynamic objects (i.e., objects whose `length` changes during execution).
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,46 @@
+{{alias}}( obj, n, predicate[, thisArg ] )
```
```suggestion

{{alias}}( obj, n, predicate[, thisArg ] )
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,46 @@
+{{alias}}( obj, n, predicate[, thisArg ] )
+    Tests whether an object's own properties 
+    satisfy a predicate function for at least `n` properties.
```
This can be improved

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,46 @@
+    - `key`: object key
+    - `obj`: the input object
+
+    The function immediately returns 
+    upon finding `n` successful properties.
```
```suggestion
    The function immediately returns  upon finding `n` successful properties.
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,46 @@
+    n: number
+        Minimum number of successful properties.
+
+    predicate: Function
+        Test function.
```
```suggestion
    predicate: Function
        The test function.
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,46 @@
+    Returns
+    -------
+    bool: boolean
+        The function returns `true` 
+        if an object's own properties satisfy a predicate 
+        for at least `n` properties; 
+        otherwise, the function returns `false`.
```
```suggestion
        The function returns `true` if an object's own properties satisfy a predicate 
        for at least `n` properties; otherwise, the function returns `false`.
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/examples/index.js:
```diff
@@ -0,0 +1,37 @@
+}
+
+var bool;
+var obj = {};
```
```suggestion
var obj;
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/examples/index.js:
```diff
@@ -0,0 +1,37 @@
+var bool;
+var obj = {};
+var i;
+
+for ( i = 0; i < 100; i++ ) {
```
```suggestion

obj = {};
for ( i = 0; i < 100; i++ ) {
```

Pranavchiku commented on lib/node_modules/@stdlib/utils/some-own-by/package.json:
```diff
@@ -0,0 +1,73 @@
+{
+    "name": "@stdlib/utils/some-own-by",
+    "version": "0.0.0",
+    "description": "Test whether some own properties of a provided object satisfies a predicate function.",
```
fix description

Pranavchiku reviewed (changes requested):
Good progress, requested a few changes, once done this PR can be reviewed.

performant23 committed:
feat(@stdlib/utils): if applied, this commit will add 'someOwnBy' in utils (Revision 1)

Add a utility to test whether some own properties of a provided object satisfies a predicate function

Signed-off-by: Pranavchiku goswami.4@iitj.ac.in

performant23 committed:
feat(@stdlib/utils): if applied, this commit will add 'someOwnBy' in utils (Revision 1)

Add a utility to test whether some own properties of a provided object satisfies a predicate function

Signed-off-by: Pranavchiku goswami.4@iitj.ac.in

performant23 commented:
Hi all, I have applied changes from code review but when I'm trying to push the code, I'm getting this:
```
git push origin feature/add-@stdlib/utils/some-own-by
Checking if remote branch exists...
Remote branch exists.
Checking for commits to push...
Found commits to push.
Checking licenses...
No dependency licensing issues detected.
Running JavaScript example files...

Running example: D:/stdlib/lib/node_modules/@stdlib/utils/some-own-by/examples/index.js
false
Running README JavaScript examples...

Running Markdown JavaScript examples: D:/stdlib/lib/node_modules/@stdlib/utils/some-own-by/README.md
lib\node_modules\@stdlib\utils\some-own-by\README.md
  1:1  error  Error: unexpected error. Encountered an error when executing code block. File: `lib\node_modules\@stdlib\utils\some-own-by\README.md`. Message: `Invalid Unicode escape sequence`.
    at next (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:259:13)
    at done (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:318:4)
    at next (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:268:12)
    at done (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:318:4)
    at next (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:155:12)
    at done (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:318:4)
    at next (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:155:12)
    at done (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:318:4)
    at next (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:155:12)
    at done (D:\stdlib\lib\node_modules\@stdlib\_tools\remark\plugins\remark-run-javascript-examples\lib\runner.js:318:4)

× 1 error
make: *** [D:/stdlib/tools/make/lib/examples/markdown_javascript.mk:89: markdown-examples-javascript-files] Error 1

Encountered an error when running README JavaScript examples.

error: failed to push some refs to 'https://github.com/performant23/stdlib.git'
```

I checked the examples included as markdown in README but I'm not sure what's causing the error (While it says invalid unicode escape sequence, the marked down examples don't use unicode esape sequences and checked the paths too). So, I'd be really grateful if you could guide me on solving this! Thank you so much!

[README.md](https://github.com/stdlib-js/stdlib/files/14469009/README.md)

performant23 committed:
feat(@stdlib/utils): revision 1

Add a utility to test whether some own properties of a provided object satisfies a predicate function

Signed-off-by: Pranavchiku goswami.4@iitj.ac.in

performant23 committed:
feat(@stdlib/utils): revision 1

Add a utility to test whether some own properties of a provided object satisfies a predicate function

Signed-off-by: Pranavchiku goswami.4@iitj.ac.in

performant23 committed:
feat(@stdlib/utils): revision 1

Add a utility to test whether some own properties of a provided object satisfies a predicate function

Signed-off-by: Pranavchiku goswami.4@iitj.ac.in

performant23 committed:
feat(@stdlib/utils): if applied, this commit will add 'someOwnBy' in utils (Revision 2)

Apply changes from code review

Signed-off-by: Pranavchiku goswami.4@iitj.ac.in

performant23 committed:
feat(@stdlib/utils): if applied, this commit will add 'someOwnBy' in utils (Revision 3)

Apply changes from code review (Except README.md)

Signed-off-by: Pranavchiku goswami.4@iitj.ac.in

performant23 committed:
Apply suggestions from code review

Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: Rutam <138517416+performant23@users.noreply.github.com>

performant23 commented:
Thank you so much for the review! I have made the required changes. Please do let me know if there are any additional changes needed!

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,45 @@
+
+{{alias}}( obj, n, predicate[, thisArg ] )
+    Tests whether some `own` properties of a provided object
+    satisfies a predicate function for at least `n` properties.
```
```suggestion
    satisfy a predicate function for at least `n` properties.
```

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,45 @@
+        Minimum number of successful properties.
+
+    predicate: Function
+        The test function.
```
```suggestion
        Test function.
```

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt:
```diff
@@ -0,0 +1,45 @@
+    Returns
+    -------
+    bool: boolean
+        The function returns `true` if an object's own properties satisfy a 
+        predicate
+        for at least `n` properties; otherwise, the function returns `false`.
```
```suggestion
        The function returns `true` if an object's own properties satisfy a 
        predicate for at least `n` properties; otherwise, the function 
        returns `false`.
```

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/package.json:
```diff
@@ -0,0 +1,73 @@
+{
+    "name": "@stdlib/utils/some-own-by",
+    "version": "0.0.0",
+    "description": "Test whether some `own` properties of a provided object satisfies a predicate function for at least `n` properties.",
```
```suggestion
    "description": "Test whether some `own` properties of a provided object satisfy a predicate function for at least `n` properties.",
```

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/package.json:
```diff
@@ -0,0 +1,73 @@
+        "object",
+        "object.keys",
+        "iterate",
+        "collection",
```
```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/lib/main.js:
```diff
@@ -0,0 +1,88 @@
+* // returns true
+*/
+function someOwnBy( obj, n, predicate, thisArg ) {
+	var count;
+	var keys;
+	var out;
+	var len;
+	var i;
+	if ( !isObject( obj ) ) {
+		throw new TypeError( format( 'invalid argument. First argument must be an object. Value: `%s`.', obj ) );
+	}
+	if ( !isPositiveInteger( n ) ) {
+		throw new TypeError( format( 'invalid argument. Second argument must be a positive integer. Value: `%s`.', n ) );
+	}
+	if ( !isFunction( predicate ) ) {
+		throw new TypeError( format( 'invalid argument. Third argument must be a function. Value: `%s`.', predicate ) );
+	}
+	keys = Object.keys( obj );
+	len = keys.length;
+	count = 0;
+	for ( i = 0; i < len; i++ ) {
+		out = predicate.call( thisArg, obj[ keys[ i ] ], keys[ i ], obj );
+		if ( out ) {
+			count += 1;
+			if ( count === n ) {
+				return true;
+			}
+		}
+		// Account for dynamically resizing an object:
+		len = keys.length;
+	}
+	return false;
```
```suggestion
	var count;
	var out;
	if ( !isObject( obj ) ) {
		throw new TypeError( format( 'invalid argument. First argument must be an object. Value: `%s`.', obj ) );
	}
	if ( !isPositiveInteger( n ) ) {
		throw new TypeError( format( 'invalid argument. Second argument must be a positive integer. Value: `%s`.', n ) );
	}
	if ( !isFunction( predicate ) ) {
		throw new TypeError( format( 'invalid argument. Third argument must be a function. Value: `%s`.', predicate ) );
	}
	count = 0;
	for ( key in obj ) {
		if ( hasOwnProp( obj, key ) ) {
		out = predicate.call( thisArg, obj[ keys[ i ] ], keys[ i ], obj );
		if ( out ) {
			count += 1;
			if ( count === n ) {
				return true;
			}
		}
	}
	return false;
```

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/lib/main.js

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/package.json

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/package.json

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/docs/repl.txt

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/lib/main.js:
```diff
@@ -0,0 +1,82 @@
+
+var isObject = require( '@stdlib/assert/is-object' );
+var isPositiveInteger = require( '@stdlib/assert/is-positive-integer' ).isPrimitive;
+var isFunction = require( '@stdlib/assert/is-function' );
+var format = require( '@stdlib/string/format' );
```
```suggestion
var hasOwnProp = require( '@stdlib/assert/has-own-property' );
var isFunction = require( '@stdlib/assert/is-function' );
var format = require( '@stdlib/string/format' );
```

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/lib/main.js

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/lib/main.js:
```diff
@@ -0,0 +1,83 @@
+		throw new TypeError( format( 'invalid argument. Third argument must be a function. Value: `%s`.', predicate ) );
+	}
+	count = 0;
+	for ( key in obj ) {
+		if ( hasOwnProp( obj, key ) ) {
+		out = predicate.call( thisArg, obj[ keys[ i ] ], keys[ i ], obj );
+		if ( out ) {
+			count += 1;
+			if ( count === n ) {
+				return true;
+			}
+		}
+	}
```
```suggestion
	for ( key in obj ) {
		if ( hasOwnProp( obj, key ) ) {
			out = predicate.call( thisArg, obj[ keys[ i ] ], keys[ i ], obj );
			if ( out ) {
				count += 1;
				if ( count === n ) {
					return true;
				}
			}
		}
	}
```

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/lib/main.js

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/lib/main.js:
```diff
@@ -0,0 +1,84 @@
+* // returns true
+*/
+function someOwnBy( obj, n, predicate, thisArg ) {
+	var count;
+	var out;
```
```suggestion
	var count;
	var out;
	var key;
```

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/lib/main.js

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented:
@performant23 Thanks; looked good! Will merge once CI passes.

(Ignore the suggestions and edits; I changed the iteration to directly loop over the object's keys without usage of `Object.keys` but broke stuff in the process that I had to fix...)

Planeshifter commented on lib/node_modules/@stdlib/utils/some-own-by/lib/main.js:
```diff
@@ -0,0 +1,85 @@
+	count = 0;
+	for ( key in obj ) {
+		if ( hasOwnProp( obj, key ) ) {
+			out = predicate.call( thisArg, obj[ keys[ i ] ], keys[ i ], obj );
```
```suggestion
			out = predicate.call( thisArg, obj[ key ], key, obj );
```

Planeshifter committed:
Update lib/node_modules/@stdlib/utils/some-own-by/lib/main.js

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
